types/node/http.d.ts

```diff
@@ -1570,6 +1570,43 @@ declare module "http" {
          * @since v0.11.4
          */
         destroy(): void;
+
+        /**
+         * Produces a socket/stream to be used for HTTP requests.
+         * By default, this function is the same as net.createConnection(). However, custom agents may override this method in case greater flexibility is desired.
+         * A socket/stream can be supplied in one of two ways: by returning the socket/stream from this function, or by passing the socket/stream to callback.
+         * This method is guaranteed to return an instance of the <net.Socket> class, a subclass of <stream.Duplex>, unless the user specifies a socket type other than <net.Socket>.
+         *
+         * @since v0.11.4
+         */
+        createConnection(options: NetConnectOpts, callback?: (err: Error | undefined, stream: stream.Duplex) => void): stream.Duplex;
+
+        /**
+         * Called when socket is detached from a request and could be persisted by the Agent.
+         * This method can be overridden by a particular Agent subclass. If this method returns a falsy value, the socket will be destroyed instead of persisting it for use with the next request.
+         *
+         * @since v8.1.0
+         */
+        keepSocketAlive(socket: stream.Duplex): void;
+
+        /**
+         * Called when socket is attached to request after being persisted because of the keep-alive options.
+         * This method can be overridden by a particular Agent subclass.
+         *
+         * @since v8.1.0
+         */
+        reuseSocket(socket: stream.Duplex, request: ClientRequest): void;
+
+        /**
+         * Get a unique name for a set of request options, to determine whether a connection can be reused.
+         * For an HTTP agent, this returns host:port:localAddress or host:port:localAddress:family. For an HTTPS agent, the name includes the CA, cert, ciphers, and other HTTPS/TLS-specific options that determine socket reusability.
+         */
+        getName(options?: {
+            host: string;
+            port: number;
+            localAddress: string;
+            family?: 4 | 6 | undefined
+        }): string;
```

Comment:
In general, @types/node abstracts out options objects as their own interfaces for clarity. This can go below the `AgentOptions` declaration:

```ts
interface AgentGetNameOptions {
    host?: string | undefined;
    port?: number | undefined;
    localAddress?: string | undefined;
    family?: 4 | 6 | undefined;
}
```

Note that all of these individual options are treated as optional within `getName()`.